Update heap setting documentation in light of machine dependent heap (#66567)
elastic/elasticsearch: docs/reference/setup/advanced-configuration.asciidoc

[[advanced-configuration]]
=== Advanced configuration settings

> 

Author: I've added some language here.

> 

Reviewer: New copy looks great to me! Thanks!

compressed oops. The exact threshold varies but 26 GB is safe on most
systems and can be as large as 30 GB on some systems. You can verify that
you are under this threshold by starting {es} with the JVM options
`-XX:+UnlockDiagnosticVMOptions -XX:+PrintCompressedOopsMode` and looking for
a line like the following:
+
[source,txt]
----
heap address: 0x000000011be00000, size: 27648 MB, zero based Compressed Oops
----
+
This line shows that zero-based compressed oops are enabled. If zero-based
compressed oops are not enabled, you'll see a line like the following instead:
+
[source,txt]
----
heap address: 0x0000000118400000, size: 28672 MB, Compressed Oops with base: 0x00000001183ff000
----

The more heap available to {es}, the more memory it can use for its internal
caches, but the less memory it leaves available for the operating system to use
for the filesystem cache. Also, larger heaps can cause longer garbage
collection pauses.

Here is an example of how to set the heap size via a `jvm.options.d/` file:

[source,txt]
------------------
-Xms2g <1>
-Xmx2g <2>
------------------
<1> Set the minimum heap size to 2g.
<2> Set the maximum heap size to 2g.

> Using `jvm.options.d` is the preferred method for configuring the heap size for
> production deployments.

Reviewer: Tried to streamline this a bit:
```suggestion
In production, we recommend using `jvm.options.d` to configure heap sizes.
```

> It is also possible to set the heap size via the `ES_JAVA_OPTS` environment
> variable. This is generally discouraged for production deployments but is useful
> for testing because it overrides all other means of setting JVM options.

Reviewer: Tried to streamline this. I'd combine with the previous para.
```suggestion
For testing, you can also set the heap sizes using the `ES_JAVA_OPTS`
environment variable. The `ES_JAVA_OPTS` variable overrides all other JVM
options. We do not recommend using `ES_JAVA_OPTS` in production.
```

[source,sh]
------------------
ES_JAVA_OPTS="-Xms2g -Xmx2g" ./bin/elasticsearch <1>
ES_JAVA_OPTS="-Xms4000m -Xmx4000m" ./bin/elasticsearch <2>
------------------
<1> Set the minimum and maximum heap size to 2 GB.
<2> Set the minimum and maximum heap size to 4000 MB.

NOTE: Configuring the heap for the <<windows-service,Windows service>> is
different than the above. The values initially populated for the Windows
service can be configured as above but are different after the service has been

> installed. Consult the <<windows-service,Windows service documentation>> for
> additional details.

Reviewer: Simplified the xref a little.
```suggestion
installed. See <<windows-service>>.
```
